### PR TITLE
feat: ✨ Implement `TrieVerifier` satisfying `CommitmentVerifier` for challenges verification

### PR DESCRIPTION
**Pull request by ffarall**

### CODE REVIEW

**Review by TDemeco — COMMENTED** [review]
Left some comments!

File: support/primitives/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        let mut next_leaf = leaves.next();
+        let mut trie_iter = match trie.iter() {
+            Ok(iter) => iter,
+            Err(_) => return Err("Failed to create trie iterator.".into()),
+        };
+        let mut challenges_iter = challenges.iter();
+
+        // Iterate over the challenges and check if there is a leaf pair of consecutive
+        // leaves that match the challenge, or an exact leaf that matches the challenge.
+        while let Some(challenge) = challenges_iter.next() {
```
[the remark this answers is not in the document]
Comment (snowmead) [reply support/primitives/src/lib.rs:54]: Done 👍🏼

File: support/primitives/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            while next_leaf.is_some()
+                && next_leaf
+                    .clone()
+                    .expect("Just checked that next_leaf is Some")
+                    .0
+                    < challenge.as_ref().to_vec()
+            {
+                prev_leaf = next_leaf.clone();
+                next_leaf = leaves.next();
+            }
```
[the remark this answers is not in the document]
Comment (snowmead) [reply support/primitives/src/lib.rs]: This has been removed in favor of double ended iteration

File: support/primitives/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                // The challenge is the leaf itself (i.e. the challenge exists in the trie).
+                (_, Some(Ok(next_leaf))) if next_leaf == challenge.as_ref().to_vec() => continue,
+                // Scenario 2 (valid): `prev_leaf` and `next_leaf` are consecutive leaves.
+                // The challenge is between the two leaves (i.e. the challenge exists in the trie).
+                (Some(Ok(prev_leaf)), Some(Ok(next_leaf)))
+                    if prev_leaf < challenge.as_ref().to_vec()
+                        && challenge.as_ref().to_vec() < next_leaf =>
+                {
+                    continue
+                }
```
Comment (TDemeco) [inline support/primitives/src/lib.rs:90]: Could you go through the logic that makes it so a proof is valid if it contains any two leaves that have the challenge between them? I'm pretty sure we talked about this but I want to make sure that it works for ANY two leaves, be it both on the same subtree, only one of them, both in different subtrees, etc

Comment (snowmead) [reply support/primitives/src/lib.rs]: I added checks to make sure `prev_leaf` and `next_leaf` are indeed leaves (i.e. they have values). So we now have certainty that `prev_leaf` and `next_leaf` are the neighbouring _leaves_ proven.

**Pull request merged**